64 bit interrupts

DrDeano opened the pull request:
Add 64 bit interrupts to handle IRQs and ISR exceptions.
Refactored the x86 architecture code for common code.
Part of #169

DrDeano commented on src/kernel/arch/x86/32bit/arch.zig:
```diff
@@ -419,12 +391,14 @@ pub fn getDateTime() DateTime {
+    gdt_common.init(gdt_32.Tss, &gdt_32.tss_entry);
+    idt_common.init(idt_32.IdtEntry, &idt_32.table);
+
+    _ = interrupts;
```
This line is needed to reference the `getInterruptStub()` function, unless there is a better way of doing it.

SamTebbs33 replied:
Comment here would be good too.

DrDeano commented on src/kernel/arch/x86/64bit/arch.zig:
```diff
@@ -332,15 +338,23 @@ pub fn initMem(boot_payload: BootPayload) Allocator.Error!MemProfile {
+    gdt_common.init(gdt_64.Tss, &gdt_64.tss_entry);
+    idt_common.init(idt_64.IdtEntry, &idt_64.table);
+
+    _ = interrupts;
```
This line is needed to reference the `getInterruptStub()` function, unless there is a better way of doing it.

SamTebbs33 replied:
I suggest adding a comment explaining that so we don't forget why it's there

SamTebbs33 commented on src/kernel/arch/x86/32bit/interrupts.zig:
```diff
@@ -1,48 +1,28 @@
         \\call  handler
         \\mov   %%eax, %%esp
+        :
+        : [offset] "rm" (gdt_common.KERNEL_DATA_OFFSET)
```
Much cleaner :)

SamTebbs33 commented on src/kernel/arch/x86/64bit/interrupts.zig:
```diff
@@ -0,0 +1,59 @@
+export fn commonStub() callconv(.Naked) void {
+    // Calling convention for x86_64 is for first parameter is in RDI (not the stack as for x86 (32 bit))
+    asm volatile (
+        \\push %%rdx
```
That's a lot of registers

DrDeano replied:
ik right, welcome to 64 bit land

SamTebbs33 reviewed (approved):
Noice

DrDeano committed:
x86_64 interrupts + refactor of common x86 structures

This creates a common interrupt functions across 32 and 64 bit versions.
This also improves the common structures across 32 and 64 bit version including:
 - GDT
 - IDT
 - ISR
 - IRQ
 - PIC
 - Arch

Update hash map for new zig version
